File: Contest-Classes/WeightedGraph.cpp
#include <bits/stdc++.h>

using namespace std;

class WeightedGraph
{
private:

    class Edge
    {
    public:

        int node;
        int cost;
        size_t next;

        Edge(int _node, int _cost, size_t _next) : node(_node), cost(_cost), next(_next){
        }
    };

    vector<Edge> graph;
    vector<int> head;
    int N;

public:

    static const int NIL = -1;

    WeightedGraph(int _N) : graph(), head(), N(_N){
        head.resize(N, NIL);
    }

    void addEdge(int x, int y, int w)
    {
        assert(1 <= x && x <= N);
        x--;

        graph.push_back(Edge(y, w, head[x]));
        head[x] = graph.size() - 1;
    }

    void addDoubleEdge(int x, int y, int w)
    {
        addEdge(x, y, w);
        addEdge(y, x, w);
    }

    int getHead(int node) const
    {
        assert(1 <= node && node <= N);
        return head[node - 1];
    }

    int getNeighbour(size_t p) const
    {
        assert(p < graph.size());
        return graph[p].node;
    }

    int getNext(size_t p) const
    {
        assert(p < graph.size());
        return graph[p].next;
    }

    int getCost(size_t p) const
    {
        assert(p < graph.size());
        return graph[p].cost;
    }
};

int main()
{
    cout << "Hello world!" << endl;
    return 0;
}

File: Java/WeightedGraph.java
import java.util.ArrayList;
import java.util.Arrays;

class WeightedGraph{
    private static class Edge{
        int node;
        int cost;
        int next;

        Edge(int node, int cost, int next){
            this.node = node;
            this.cost = cost;
            this.next = next;
        }
    }

    final static int NIL = -1;
    private int[] head;
    private ArrayList<Edge> graph;

    private int N;
    private int counter;

    WeightedGraph(int N){
        initialize(N);
    }

    public int getN() {
        return N;
    }

    private void initialize(final int N){
        head = new int[N];
        graph = new ArrayList<>();

        this.N = N;
        this.counter = 0;

        Arrays.fill(head, NIL);
    }

    void addEdge(int x, int y, int w){
        if (!(1 <= x && x <= N)) throw new AssertionError();
        x--; y--;
        graph.add(new Edge(y, w, head[x]));
        head[x] = counter++;
    }

    int getHead(int node){
        if (!(1 <= node && node <= N)) throw new AssertionError();
        node--;
        return head[node];
    }

    int getNext(int p){
        if (!(0 <= p && p < counter)) throw new AssertionError();
        return graph.get(p).next;
    }

    int getNeighbour(int p){
        if (!(0 <= p && p < counter)) throw new AssertionError();
        return graph.get(p).node + 1;
    }

    int getCost(int p){
        if (!(0 <= p && p < counter)) throw new AssertionError();
        return graph.get(p).cost;
    }
    
    WeightedGraph transpose(){
        WeightedGraph GT = new WeightedGraph(N);

        for (int i = 1; i <= N; i++) {
            for (int son : getNeighbours(i))
                GT.addEdge(son, i);
        }

        return GT;
    }

    List<Integer> getNeighbours(int node){
        if (!(1 <= node && node <= N)) throw new AssertionError();

        List<Integer> list = new ArrayList<>();

        for (int p = head[node - 1]; p != NIL; p = graph.get(p).next) {
            list.add(graph.get(p).node + 1);
        }

        return list;
    }
}
